include/El/optimization/bfgs.hpp

```diff
@@ -0,0 +1,212 @@
+/*
+   Copyright (c) 2009-2016, Ryan H. Lewis
+   All rights reserved.
+
+   This file is part of Elemental and is under the BSD 2-Clause License, 
+   which can be found in the LICENSE file in the root directory, or at 
+   http://opensource.org/licenses/BSD-2-Clause
+*/
+#ifndef EL_OPTIMIZATION_BFGS_HPP
+#define EL_OPTIMIZATION_BFGS_HPP
+#include <utility>
+#include <El.hpp>
+
+namespace El {
+
+template< typename T, typename Function, typename Gradient>
+T zoom( const Function& f, const Gradient& gradient, T f0, const DistMatrix<T>& x0, const DistMatrix<T>& p,
+        T alpha_low, T alpha_high, T c1, T c2){
+    DistMatrix<T> x_j(x0);
+    DistMatrix<T> g2(p.Height(), 1);
+    while (alpha_high - alpha_low > 10*El::limits::Epsilon<T>()){
+        T alpha = (alpha_low + alpha_high)/2.0;
```

Comment:
I would highly recommend against ever using floating-point literals in templated routines and would instead recommend `T(2)` instead of `T(2.0)`. This happens to be safe for `2.0` but the conversion generally leads to a loss in precision.